### PR TITLE
Optimize Runtime Perf

### PR DESCRIPTION
**Pull request by AndreSlavescu**
## Summary

Optimizing Softmax and RMSNorm runtime performance on hidden_size >= 64k

## Testing Done

Added large tests for 64K dim

- Hardware Type: RTX 3090
- [x] run `make test` to ensure correctness
- [x] run `make checkstyle` to ensure code style
- [x] run `make test-convergence` to ensure convergence

### CODE REVIEW

File: src/liger_kernel/ops/rms_norm.py [hunk 1]
```diff
@@ -554,12 +560,14 @@ def forward(ctx, X, W, eps, offset=0.0, casting_mode="llama", in_place=True, row
         W: (H,)
         """
         Y, X, RSTD, BLOCK_SIZE, num_warps, casting_mode = rms_norm_forward(X, W, eps, offset, casting_mode, row_mode)
+        num_stages = calculate_num_stages()
```
Comment (Tcc0403) [inline src/liger_kernel/ops/rms_norm.py:563]: nit: We can just return `num_stages` from `rms_norm_forward()` like `num_warps` to avoid calling it again.

File: src/liger_kernel/ops/utils.py [hunk 2]
```diff
@@ -42,26 +42,184 @@ def maybe_to_contiguous(x):
...
+        "num_sm": torch_device_props.multi_processor_count,
+        "max_shared_mem": max_shared_mem,
+        "max_threads_per_sm": torch_device_props.max_threads_per_multi_processor,
+        "warp_size": 64 if is_hip_device else 32,
+        "max_regs": torch_device_props.regs_per_multiprocessor,
+        "is_hip": is_hip_device,
+    }
+
+
+def calculate_num_stages():
```
Comment (Tcc0403) [inline src/liger_kernel/ops/utils.py:153]: Is there a table where we can look up these properties?

File: test/transformers/test_softmax.py [hunk 3]
```diff
@@ -101,3 +101,42 @@ def test_liger_softmax_functional(shape, dtype, atol, rtol):
...
+            5e-2,
+            5e-2,
+            marks=pytest.mark.skipif(
+                not supports_bfloat16(),
+                reason="bfloat16 not supported on this device",
+            ),
+        ),
+    ],
+)
+def test_large_64k_softmax_correctness(dtype, atol, rtol):
```
Comment (Tcc0403) [inline test/transformers/test_softmax.py:121]: Is there any considerations of not just adding a test case to original tests?

File: test/transformers/test_rms_norm.py [hunk 4]
```diff
@@ -183,3 +183,49 @@ def test_correctness_functional(bs, sl, hd, dtype, atol, rtol, reference, offset
...
+    ],
+)
+@pytest.mark.parametrize(
+    "reference, offset, casting_mode",
+    [
+        (LlamaRMSNorm, 0.0, "llama"),
+        (GemmaRMSNorm, 1.0, "gemma"),
+    ],
+)
+def test_large_64k_correctness(dtype, atol, rtol, reference, offset, casting_mode):
```
Comment (Tcc0403) [inline test/transformers/test_rms_norm.py:207]: ditto

File: src/liger_kernel/ops/utils.py [hunk 5]
```diff
@@ -42,26 +42,184 @@ def maybe_to_contiguous(x):
...
+
+    # Triton requires num_warps to be a power of 2.
+    num_warps = _prev_power_of_2(num_warps)
+
     return BLOCK_SIZE, num_warps
 
 
+def _get_device_properties():
+    device = torch.cuda.current_device()
+    torch_device_props = torch.cuda.get_device_properties(device)
```
Comment (Tcc0403) [inline src/liger_kernel/ops/utils.py:131]: We should make it xpu compatible
https://docs.pytorch.org/docs/stable/generated/torch.xpu.get_device_properties.html#torch.xpu.get_device_properties

File: src/liger_kernel/ops/utils.py [hunk 6]
```diff
@@ -42,26 +42,184 @@ def maybe_to_contiguous(x):
...
+    MAX_FUSED_SIZE = _calculate_max_fused_size()
     BLOCK_SIZE = triton.next_power_of_2(n)
     if BLOCK_SIZE > MAX_FUSED_SIZE:
         raise RuntimeError(
             f"Cannot launch Triton kernel since n = {n} exceeds the recommended Triton blocksize = {MAX_FUSED_SIZE}."
         )
 
     num_warps = 4
-    if BLOCK_SIZE >= 32768:
+    if BLOCK_SIZE >= 65536:
```
Comment (Tcc0403) [inline src/liger_kernel/ops/utils.py:107]: By the way, I'm always wondering why we don't take element_size into account. Do you have any idea?

**Comment by shimizust** [comment]
Can you show some plots comparing the perf before/after?
